src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -61,57 +60,72 @@ func NewServiceConfigFetcher(mf *metadata.MetadataFetcher, opts options.ConfigGe
 			Timeout: opts.HttpRequestTimeout,
 		},
 		serviceName: serviceName,
-		mf:          mf,
 		opts:        opts,
+		accessToken: accessToken,
+	}
+
+	configIdFetcher := NewServiceConfigIdFetcher(serviceName, opts.ServiceControlURL,
+		s.client, accessToken)
+	s.newConfigId = func() (string, error) {
+		return configIdFetcher.fetchNewConfigId()
 	}
-	return scf, nil
+
+	return s, nil
 }
 
 // Fetch the service config by given configId. If configId is empty, try to
-// fetch the latest service config,.
-func (scf *ServiceConfigFetcher) FetchConfig(configId string) (*confpb.Service, error) {
+// fetch the latest service config.
+func (s *ServiceConfigFetcher) FetchConfig(configId string) (*confpb.Service, error) {
 	_fetchConfig := func(configId string) (*confpb.Service, error) {
+
 		if configId != "" {
-			token, _, err := scf.accessToken()
+			if configId == s.curConfigId() {
+				return nil, nil
+			}
+
+			token, _, err := s.accessToken()
 			if err != nil {
 				return nil, fmt.Errorf("fail to get access token: %v", err)
 			}
-			return scf.callServiceManagement(util.FetchConfigURL(scf.opts.ServiceManagementURL, scf.serviceName, configId), token)
+
+			return s.callServiceManagement(util.FetchConfigURL(s.opts.ServiceManagementURL, s.serviceName, configId), token)
 		}
 
-		glog.Infof("check new rollouts for service %v", scf.serviceName)
-		newRolloutId, newConfigId, err := scf.loadConfigFromRollouts(scf.serviceName, scf.curRolloutId, scf.curConfigId())
+		glog.Infof("check new config id for service %v", s.serviceName)
+		newConfigId, err := s.newConfigId()
```

Comment:
Did not quite follow the logic here. I could not find where newConfigId() is defined.

Reply:
Basically, what we need to do is:
1) on very minute timer, we call empty report to get latest Rollout ID. 
2) if rollout ID is changed,  call ServiceManaement to get Latest rollout info.   A rollout may have multiple service config_id.  ESPv1, we support traffic percentage rollout.  But ESPv2, we don't.  We just get the first largest pecentage config_id.  That is the config_id we need to use.
E.g.  a rollout may have 80% for id1 and 20% for id2.  we just use id1 since it is 80%.

3) use that config_id to get the service config.